cdap-ui/app/cdap/services/ThemeHelper.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+
+import cssVars from 'css-vars-ponyfill';
+import {objectQuery, isNilOrEmpty} from 'services/helpers';
+
+declare global {
+  interface Window {
+    CDAP_UI_THEME: any;
+  }
+}
+
+export function applyTheme() {
+  if (
+    !objectQuery(window, 'CDAP_UI_THEME', 'styles') ||
+    isNilOrEmpty(window.CDAP_UI_THEME.styles)
+  ) {
+    // need to run this at least once even if there's no theme customization
+    // so that css variables are parsed correctly even in older browsers
+    cssVars();
+    return;
+  }
+
+  // this is what's going on under the hood for modern browsers:
+  // document.documentElement.style.setProperty(`--${cssVar}`, cssValue);
+  cssVars({
+    variables: window.CDAP_UI_THEME.styles,
+  });
+}
+
+interface ThemeObj {
+  [key: string]: any;
+}
+
+function getTheme(): ThemeObj {
+  if (isNilOrEmpty(window.CDAP_UI_THEME)) {
+    return {};
+  }
+
+  let theme: ThemeObj = {};
+  const themeJSON = window.CDAP_UI_THEME;
+  const specVersion = themeJSON['spec-version'];
+
+  if (specVersion === '1.0') {
+    theme = get10Content(theme, themeJSON);
+    theme = get10Features(theme, themeJSON);
```

Comment:
this is quite weird, you are passing the theme as function argument, and you are modifying the object in the function. Then you assign the return value back to the object as well. As much as possible, avoid modifying function arguments.

Also, might be better to have a single parser function that will give the final object. Inside that single function, you can break it down to internal contents and features functions.